feat: add dynamic onboarding

kiritkir opened the pull request:
#20

kiritkir commented:
@JohnTrunix:

What I've done:
- fetched repository URLs via PyPI registry (its not hardcoded)
- used `packaging.version` instead of regex to match Git tags 
- cloned the matching tag using `GitPython`
- Left CLI integration out for now as discussed
- Tested it using a separate `pyproject_test.toml` file (not comitted -can add it if helpful)

What I'll be working on next:
- build a js equivalent that parses `package.json` dependencies
- extend metadata resolution (error handling)

Would love to hear if this looks like the right direction to you. Just want to make sure I’m aligned before moving forward 🚀

JohnTrunix commented:
@kiritkir , looks very promising! Nice approach, keep it up 🚀

JohnTrunix commented on .gitignore:
```diff
@@ -13,3 +13,5 @@ dist/
 coverage.xml
 junit.xml
 pylint_results.xml
+**/.DS_Store
```
just ignore .DS_Store this will then ignore all such files everywhere in the project

JohnTrunix commented on .gitignore:
```diff
@@ -13,3 +13,5 @@ dist/
 junit.xml
 pylint_results.xml
+**/.DS_Store
+.gitignore
```
What? Dont ignore the .gitignore!

kiritkir commented:
@JohnTrunix:

i've completed everything described above.

One thing I would still like to improve:
for some packages like `torch`, the script returns: "No GitHub repo found for torch" even though we know the code is on GitHub e.g. https://github.com/pytorch/pytorch. 
The issue seems to be that the PyPI registry metadata doesnt always include a clean or structured repository field, which makes it tricky to extract. The same goes for Tag but there it will say "N/A" if there is none.

As a next step i'll try to find a clean way to handle this (maybe a fallback logic or heuristics idk)

Let me know if that makes sense of if you'd like me to prioritize smt else :)

JohnTrunix commented on sandbox/dep_resolver.py:
rename the file to xxx_resolver.py which indicates that it is resolving pyhton deps

kiritkir replied:
done

JohnTrunix replied:
dep is too general, use something like py_resolver.py, I saw you have already a new file, so then delete the old one ;)

JohnTrunix commented on sandbox/dep_resolver.py:
```diff
@@ -0,0 +1,150 @@
+
+
+if __name__ == "__main__":
+    process(Path("..") / "pyproject.toml")
```
replace with `process(Path().cwd().parent / "pyproject.toml")` or like you did for js deps, just create a pyproject.toml with some example deps in the sandbox folder ;)

JohnTrunix replied:
?

JohnTrunix commented on sandbox/dep_resolver.py:
```diff
@@ -0,0 +1,150 @@
+from packaging.version import parse as parse_version
+
+
+def get_pyproject_dependencies(pyproject_path: Path) -> dict:
```
Instead of manually parsing, an evaluation of this might be worth https://stackoverflow.com/a/76143344 but I am not sure if this is better... but you already use this lib for other stuff so its might worth a look

JohnTrunix commented on sandbox/dep_resolver.py:
```diff
@@ -0,0 +1,150 @@
+) -> tuple[str, Version] | None:
+    matching_versions = []
+    for tag in tags:
+        cleaned = tag.lstrip("vV-").replace("version", "").strip("-")
```
I am not sure if a regex would suit here better. A regex something like:
- `^\d+\.\d+\.\d+$`
- `^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)$`

If you are going to use a regex make sure to test all possible matches with a tool like https://regex101.com/

JohnTrunix replied:
?

JohnTrunix commented on sandbox/npm_resolver.py:
Cool! Unfortunately not that many matches with repo tags... I dont know if we have choosen here the wrong packages for that or if this in general the case but I would just stick to this and evaluate later more...

JohnTrunix commented on .DS_Store:
REMOVE

kiritkir replied:
done

JohnTrunix commented:
@kiritkir whats the status here?

kiritkir commented:
@JohnTrunix 

I've addressed all the requested changes above. Everything is implemented and working as expected. The only open point is the `npm_resolver`: it's working but there arent many matches with repo tags.

JohnTrunix commented:
> @JohnTrunix
> 
> I've addressed all the requested changes above. Everything is implemented and working as expected. The only open point is the `npm_resolver`: it's working but there arent many matches with repo tags.

So ready for merge?

JohnTrunix commented on sandbox/pyproject_test.toml:
```diff
@@ -0,0 +1,7 @@
+version = "0.1.0"
+dependencies = [
+  "pydantic ==1.10.7",
+  "typer ==0.9.0"
```
I would use following deps: pydantic, flask, flask-cors, sqlalchemy

JohnTrunix commented:
@kiritkir if you are going to rename a file, dont copy paste the file and keep the old one, just rename the existing file, fix the requested chanes and push it. No need for tracking an old outdated file.

kiritkir commented:
@JohnTrunix : you can expect the updated version by tonight :)

kiritkir commented:
@JohnTrunix 

- I renamed `dep_resolver.py` to `py_resolver.py`
- I tried to enhance the tag detection: It now also attempts to match the PyPI specifiers with GitHub tags. It also falls back to the latest available tag if no exact match is found.
- i also added a fallback GitHub repo search using the GitHub API if not found in PyPI metadata
- also added the following deps from above in the toml file

Let me know if you would like me to change things and thanks :)

JohnTrunix commented:
@kiritkir looks good from my side, lets merge so that we can integrate it into the cli

JohnTrunix commented:
Closing this as this branch got disconnected from main, instead we use #43
